File: spark-scala-maven/src/main/scala/adj/spark/df/RDD2DataFrameReflection.scala
package adj.spark.df

import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import org.apache.spark.sql.SQLContext
import com.alibaba.fastjson.JSON
import adj.spark.df.pojo.Student

/**
 * 基于反射RDD到DataFrame转换, 
 * 必须用object extends App方式, 不能用def main运行程序, 否则会报no typetag for ... 错误
 */
object RDD2DataFrameReflection extends App{
    val conf = new SparkConf()
                 .setAppName("RDD2DataFrameReflection")
                 .setMaster("local")
    val context     = new SparkContext(conf)
    val sqlContext  = new SQLContext(context)
    
    val lineRDD = context.textFile("hdfs://master:9000/user/spark/data/students.json", 3)
    
    val studentRDD = lineRDD.map(line => {
      val js   = JSON.parseObject(line)
      
      val id   = js.getInteger("id")
      val name = js.getString("name")
      val age  = js.getInteger("age")
      
      Student(id, name, age)
    })
    
    // 隐式转换
    import sqlContext.implicits._
    val studentDF = studentRDD.toDF()
    
    studentDF.registerTempTable("students")
    
    val sqlStudentDF = sqlContext.sql("select id, name, age from students")
    sqlStudentDF.show()
    
    val studentResultRDD = sqlStudentDF.rdd.map(row => {
      val id   = row.getAs[Int]("id")
      val name = row.getAs[String]("name")
      val age  = row.getInt(2)
      
      Student(id, name, age)
    })
    
    val studentResultRDD2 = sqlStudentDF.rdd.map(row => {
      val rowMap = row.getValuesMap[Any](Array("id", "name", "age"))
      val id   = rowMap("id").toString().toInt
      val name = "_" + rowMap("name").toString()
      val age  = rowMap("age").toString().toInt
      
      Student(id, name, age)
    })
    
    studentResultRDD.foreach(println)
    studentResultRDD2.foreach(println)
}

File: spark-java/src/main/java/adj/spark/df/RDD2DataFrameReflection.java
package adj.spark.df;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.function.VoidFunction;
import org.apache.spark.sql.DataFrame;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SQLContext;

import com.alibaba.fastjson.JSONObject;

import adj.spark.df.pojo.Student;

/**
 * 使用反射的方式将RDD转换为DataFrame
 * @author felix
 */
public class RDD2DataFrameReflection {
	public static void main(String[] args) {
		SparkConf conf = new SparkConf()
							 .setMaster("local")
							 .setAppName("RDD2DataFrameReflection");
		
		JavaSparkContext context = new JavaSparkContext(conf);
		SQLContext       sqlContext = new SQLContext(context);
		
		JavaRDD<String> lineRDD = context.textFile("hdfs://master:9000/user/spark/data/students.json");
		JavaRDD<Student> studentRDD = lineRDD.map(new Function<String, Student>() {
			private static final long serialVersionUID = 1L;
			
			@Override
			public Student call(String line) throws Exception {
				JSONObject json = JSONObject.parseObject(line);
				int id = json.getInteger("id");
				String name = json.getString("name");
				int age = json.getInteger("age");
				
				return new Student(id, name, age);
			}
		});
		
		/**
		 * 使用反射方式将RDD转转为DataFrame。
		 * 注意：Student必须实现Serializable接口,并且设置field的get/set方法
		 */
		DataFrame studentDataFrame = sqlContext.createDataFrame(studentRDD, Student.class);
		
		// 拿到一个DataFrame之后,就可以将其注册为一个临时表,然后针对其中的数据执行SQL语句
		studentDataFrame.registerTempTable("students");
		
		/**
		 * 针对students临时表,执行SQL语句。
		 * 注意：如果没有指明查询字段,查询的结果是乱序的
		 */
		DataFrame sqlDataFrame = sqlContext.sql("select id,name,age from students where age < 19");
		
		// 显示执行的结果
		sqlDataFrame.show();
		
		// 将DataFrame转换为RDD
		JavaRDD<Row> sqlRowRDD = sqlDataFrame.javaRDD();
		
		JavaRDD<Student> sqlStudentRDD = sqlRowRDD.map(new Function<Row, Student>() {
			private static final long serialVersionUID = 1L;
			
			@Override
			public Student call(Row row) throws Exception {
				int id = row.getInt(0);
				String name = "_" + row.getString(1);
				int age = row.getInt(2);
				return new Student(id, name, age);
			}
		});
		
		sqlStudentRDD.foreach(new VoidFunction<Student>() {
			private static final long serialVersionUID = 1L;

			@Override
			public void call(Student t) throws Exception {
				System.out.println(t);
			}
		});
	}
}
